### ISSUE TITLE
AF allocating mem on 1st device whatever af::setDevice(X) 

### ISSUE THREAD

**Issue by WilliamTambellini** [arrayfire/arrayfire#1915]
Hi all
I think I ve found either an undesired behavior or a bug using at least AF 3.4.2:
- modified the unified/basic.cpp example that way: (commenting out the CPU and OpenCL blocks)

```
try {
	printf("Trying CUDA Backend\n");
        af::setDevice(1); // be sure to have at least 2 GPUs on your machine
        af::setBackend(AF_BACKEND_CUDA);
        testBackend();
    } catch (af::exception& e) {
        printf("Caught exception when trying CUDA backend\n");
        fprintf(stderr, "%s\n", e.what());
    }
```

- compiled as usual
cmake .. -DArrayFire_DIR=/home/wtambellini/Downloads/af/arrayfire-3.4.2/share/ArrayFire/cmake/
make -j8
- run the unified basic example targeting cuda while keeping an eye on nvidia-smi (watch -n1 nvidia-smi):
 LD_LIBRARY_PATH=/home/wtambellini/Downloads/af/arrayfire-3.4.2/lib/:/usr/local/cuda/nvvm/lib64 unified/basic_unified
- see that about 200 MB of mem are allocated on both GPUs (see screenshot)
![screenshot from 2017-08-31 12-42-18](https://user-images.githubusercontent.com/109458/29942542-1f760106-8e4b-11e7-83ca-fddc5dbf0ec4.png)

- expected behavior: mem to be allocated only on the selected device (the 2nd GPU)
Note: processing/computation seems to occur correctly on the desired device.
@umar456 : could you please have a look ?
Kind
W.

**Comment by WilliamTambellini** [comment]
I suppose playing with 
CUDA_VISIBLE_DEVICES 
would do the job but I would prefer setDevice(x) to be consistent.

**Comment by pavanky** [comment]
@WilliamTambellini The memory being used is from CUDA context that is created. ArrayFire itself does not create any memory.

You can use `AF_CUDA_DEFAULT_DEVICE=x ./my_program` so that this is not an issue.  More configuration options can be seen here: http://arrayfire.org/docs/configuring_environment.htm

**Comment by WilliamTambellini** [comment]
@pavanky  I just tried with AF_CUDA_DEFAULT_DEVICE=x and I still see mem allocated on both GPUs.
This behavior is wasting about 200MB of mem on the 1st GPU for each started process even if these are set to use other GPUs with af::setDevice(x). 
Closing the issue immediately does not seem legitimate to me: I would bet that the expected behavior of af::setDevice(x) is to not allocate any mem on any other gpu than gpu x. 
Kind regards

**Comment by pavanky** [comment]
@WilliamTambellini This should only be happening if `cudaGetDeviceProperties` is creating a context on each gpu. I don't see any other code that may be problematic. This function call is needed for defaulting to the "best" gpu, but if this is indeed the issue and you don't need sorting behavior, this can probably be disabled using an environment flag.

**Comment by WilliamTambellini** [comment]
Hi @pavanky 
Well perhaps are you looping and calling cudaGetDeviceProperties() on all devices allocating 200MB on all GPUs just for the context. 
I suppose this would be acceptable if temporary ie at initialization but as today these 200MB are lost if the process only use 1 GPU.  
Why dont you simply delete the context untill setDevice(x) has been called or only if at least 1 array is created on a device ?
Cheers
W.

**Comment by pavanky** [comment]
@WilliamTambellini "Deleting the context" (calling cudaDeviceReset()) would be fine if arrayfire was the only CUDA library being run in that application. If the app is using any other CUDA library before calling arrayfire, resetting will mess up the state.

**Comment by WilliamTambellini** [comment]
I have found a dirty workaround (in my code) in order for AF NOT to allocate any mem on any other device: 
      setenv("CUDA_VISIBLE_DEVICES", std::to_string(deviceId).c_str(), 1);

I m not knowledgeable enough in cuda to fix that properly.
Let's at least confirm the purpose of this bugfix/enhancement : 
could you (Umar & Pavan) just acknowledge this request is legitimate : 
a call to af::setDevice(x) does not have to permanently generate any mem allocation on any other device than device x ?
Cheers
W.

**Comment by pavanky** [comment]
@WilliamTambellini I already reopened the issue. @umar456 should look into it. I don't have access to multi gpu machines.

**Comment by umar456** [comment]
This is an issue in 3.4.2 because ArrayFire created a stream in the setActiveDevice function. This has been resolved in the newer version of ArrayFire.

**Comment by pavanky** [comment]
@umar456 If you still plan to backport fixes to 3.4, I don't think this particular change can be easily cherry-picked into 3.4 branch. It may require a separate fix.
